File: src/lib/shared/types/Album.ts
import type {ExternalUrls} from "./ExternalUrls";
import type {Image} from "./Image";
import type {Restrictions} from "./Restrictions";

export interface Album {
  album_type?: "album" | "single" | "compilation";
  total_tracks: number
  available_markets?: string[] | undefined
  external_urls: ExternalUrls
  href: string
  id: string
  images?: Image[]
  label?: string | undefined
  name: string
  release_date: string
  release_date_precision: "year" | "month" | "day"
  restrictions?: Restrictions | undefined
  type: "album"
  artists: {id: string, name: string}[]
  tracks: { id: string, name: string }[]
  genres?: string[] | undefined
  popularity?: number | undefined
  uri: string;
}

export interface LibraryAlbum{
  album: {
    id: string,
    name: string,
    artists: {id: string, name: string}[]
    total_tracks: number
    images?: Image[]
  },
  added_at: string
}

export function toAlbum(album: SpotifyApi.SavedAlbumObject): Album {
  return {
    album_type: album.album.album_type,
    total_tracks: album.album.total_tracks,
    available_markets: album.album.available_markets,
    external_urls: album.album.external_urls,
    href: album.album.href,
    id: album.album.id,
    images: album.album.images,
    label: album.album.label,
    name: album.album.name,
    release_date: album.album.release_date,
    release_date_precision: album.album.release_date_precision,
    restrictions: album.album.restrictions,
    type: album.album.type,
    artists: album.album.artists.map(a => ({id: a.id, name: a.name})),
    tracks: [],
    genres: album.album.genres,
    popularity: album.album.popularity,
    uri: album.album.uri
  }
}

export function toLibraryAlbum(album: SpotifyApi.SavedAlbumObject): LibraryAlbum {
  return {
    album: {
      id: album.album.id,
      name: album.album.name,
      artists: album.album.artists.map(a => ({id: a.id, name: a.name})),
      total_tracks: album.album.total_tracks,
      images: album.album.images
    },
    added_at: album.added_at
  }
}

File: src/main/java/de/wittenbude/exportify/album/api/Album.java
package de.wittenbude.exportify.album.api;

import com.neovisionaries.i18n.CountryCode;
import jakarta.persistence.Embeddable;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.proxy.HibernateProxy;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.*;

@Getter
@Setter
@Accessors(chain = true)
@Entity
public class Album {

    @Id
    @GeneratedValue
    private UUID id;

    @CreationTimestamp
    private Instant versionTimestamp;

    private String albumType;
    private Integer totalTracks;
    private String href;
    private String spotifyID;
    private String name;
    private String releaseDate;
    private ReleaseDatePrecision releaseDatePrecision;
    private String restrictions;
    private String spotifyObjectType;
    private String uri;
    private ExternalIDs externalIDs;
    private String label;
    private Integer popularity;

    @JdbcTypeCode(SqlTypes.JSON)
    private Collection<Copyright> copyrights;

    @JdbcTypeCode(SqlTypes.JSON)
    private Collection<Image> images;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private Collection<CountryCode> availableMarkets;

    @JdbcTypeCode(SqlTypes.JSON)
    private Map<String, String> externalUrls;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private Collection<String> genres;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> spotifyArtistIDs;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> spotifyTrackIDs;

    @Override
    public final boolean equals(Object o) {
        if (this == o) return true;
        if (o == null) return false;
        Class<?> oEffectiveClass = o instanceof HibernateProxy ? ((HibernateProxy) o).getHibernateLazyInitializer().getPersistentClass() : o.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass() : this.getClass();
        if (thisEffectiveClass != oEffectiveClass) return false;
        Album album = (Album) o;
        return getId() != null && Objects.equals(getId(), album.getId());
    }

    @Override
    public final int hashCode() {
        return this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass().hashCode() : getClass().hashCode();
    }

    @Getter
    @AllArgsConstructor
    public enum ReleaseDatePrecision {
        DAY,
        MONTH,
        YEAR
    }

    @Getter
    @Setter
    @Accessors(chain = true)
    @Embeddable
    @EqualsAndHashCode
    public static class ExternalIDs {
        private String isrc;
        private String ean;
        private String upc;
    }

    @Getter
    @Setter
    @Accessors(chain = true)
    @EqualsAndHashCode
    public static class Copyright {
        private String text;
        private String type;
    }

    @Getter
    @Setter
    @Accessors(chain = true)
    @EqualsAndHashCode
    public static class Image {

        private String url;
        private Integer height;
        private Integer width;

    }
}

